### PR TITLE
Fix edge case empty waiting list

### PR DESCRIPTION
**Pull request by mariusmihaic**
## Reasoning behind the pull request
- With stakingV4, shuffled out nodes from eligible are sent to auction, instead of waiting list. This causes nodes to spend one epoch in auction before they are distributed (if selected based on topUP) to waiting. However, in case we have: `nodes in waiting < max waiting list size`, after shuffling out nodes, there is no need to send nodes to auction, since they will be selected anyway. They can be directly sent to waiting list after shuffling, because next epoch they will be selected anyway. This also fixes some edge cases when waiting lists are almost empty and nodes would be forced to spend extra epochs(auction+waiting) before being sent to eligible.
  
## Proposed changes
- In case number of active nodes <= max number of nodes, distribute all shuffled out nodes to waiting instead of auction

## Testing procedure
- Integration test for edge cases.

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: vm/systemSmartContracts/delegation.go [hunk 1]
```diff
@@ -2883,22 +2883,6 @@ func (d *delegation) executeStakeAndUpdateStatus(
 	return vmcommon.Ok
 }
 
-func (d *delegation) getConfigStatusAndGlobalFund() (*DelegationConfig, *DelegationContractStatus, *GlobalFundData, error) {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply vm/systemSmartContracts/delegation.go:2886]: Not used, remained here from a rc->feat merge

File: integrationTests/vm/staking/stakingV4_test.go [hunk 2]
```diff
@@ -901,3 +901,284 @@ func TestStakingV4_JailAndUnJailNodes(t *testing.T) {
...
+	// Since activeNodes == maxNumNodes, shuffled out nodes will always be sent directly to waiting list, instead of auction
+	epoch = 10
+	require.Equal(t, epoch, node.EpochStartTrigger.Epoch())
+	prevNodesConfig = currNodesConfig
+	for epoch < 13 {
+		node.Process(t, 5)
+
+		currNodesConfig = node.NodesConfig
+		require.Len(t, getAllPubKeys(currNodesConfig.eligible), 8)
+		require.Len(t, getAllPubKeys(currNodesConfig.waiting), 4)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply integrationTests/vm/staking/stakingV4_test.go:1049]: Right, should've done this before.
Refactored to reuse functionality with new functions

File: vm/systemSmartContracts/delegation.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply vm/systemSmartContracts/delegation.go:2886]: Not used, I think it was some merging "leftovers" from an rc->feat

File: sharding/nodesCoordinator/hashValidatorShuffler.go [hunk 3]
```diff
@@ -283,6 +285,30 @@ func shuffleNodes(arg shuffleNodesArg) (*ResUpdateNodes, error) {
 
 	shuffledOutMap, newEligible := shuffleOutNodes(newEligible, numToRemove, arg.randomness)
 
+	numShuffled := getNumPubKeys(shuffledOutMap)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply sharding/nodesCoordinator/hashValidatorShuffler.go]: Added separate function with little refactor

File: sharding/nodesCoordinator/hashValidatorShuffler.go [hunk 4]
```diff
@@ -294,12 +320,27 @@ func shuffleNodes(arg shuffleNodesArg) (*ResUpdateNodes, error) {
...
 	}
+
+	if distributeShuffledToWaitingInStakingV4 {
+		log.Debug("distributing shuffled out nodes to waiting in staking V4",
+			"num shuffled nodes", numShuffled, "num waiting nodes", numNewWaiting)
+
+		// Distribute validators from SHUFFLED OUT -> WAITING
+		err = arg.distributor.DistributeValidators(newWaiting, shuffledOutMap, arg.randomness, arg.flagBalanceWaitingLists)
+		if err != nil {
+			log.Warn("distributeValidators shuffledOut failed", "error", err)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply sharding/nodesCoordinator/hashValidatorShuffler.go]: Did some digging + tests well regarding this and it seems like we could safely return error here and it should also be backwards compatible.
Refactored to return error

**Review by sasurobert — APPROVED** [review]
We should really continue refactoring shuffleNodes function into a full fedged component.

File: sharding/nodesCoordinator/hashValidatorShuffler.go [hunk 5]
```diff
@@ -285,26 +297,42 @@ func shuffleNodes(arg shuffleNodesArg) (*ResUpdateNodes, error) {
...
+
+	if shouldDistributeShuffledToWaiting(shuffledNodesCfg) {
+		log.Debug("distributing shuffled out nodes to waiting",
+			"num shuffled nodes", shuffledNodesCfg.numShuffled, "num waiting nodes", shuffledNodesCfg.numNewWaiting)
+
 		// Distribute validators from SHUFFLED OUT -> WAITING
 		err = arg.distributor.DistributeValidators(newWaiting, shuffledOutMap, arg.randomness, arg.flagBalanceWaitingLists)
 		if err != nil {
-			log.Warn("distributeValidators shuffledOut failed", "error", err)
+			return nil, fmt.Errorf("distributeValidators shuffled out failed, error: %w", err)
```
Comment (raduchis) [inline sharding/nodesCoordinator/hashValidatorShuffler.go:335]: also backwards incompatibility here?

File: sharding/nodesCoordinator/hashValidatorShuffler.go [hunk 6]
```diff
@@ -285,26 +297,42 @@ func shuffleNodes(arg shuffleNodesArg) (*ResUpdateNodes, error) {
 
 	err = moveMaxNumNodesToMap(newEligible, newWaiting, arg.nodesMeta, arg.nodesPerShard)
 	if err != nil {
-		log.Warn("moveNodesToMap failed", "error", err)
+		return nil, fmt.Errorf("moveNodesToMap failed, error: %w", err)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply sharding/nodesCoordinator/hashValidatorShuffler.go:300]: This is related to my comment from [here](https://github.com/multiversx/mx-chain-go/pull/5061#discussion_r1134060699).
An error here could never happend, that's why it was only treated as a warn;
This case never manifested on mainnet anyway, it's safe to refactor it

File: sharding/nodesCoordinator/hashValidatorShuffler.go (hunk 5 above)
Comment (mariusmihaic) [reply sharding/nodesCoordinator/hashValidatorShuffler.go:335]: This is related to my comment from [here](https://github.com/multiversx/mx-chain-go/pull/5061#discussion_r1134060699).
An error here could never happend, that's why it was only treated as a warn;
This case never manifested on mainnet anyway, it's safe to refactor it
